src/Npgsql/NpgsqlReadBuffer.cs

```diff
@@ -337,7 +337,9 @@ public double ReadDouble(bool littleEndian)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private T Read<T>()
         {
-            Debug.Assert(Unsafe.SizeOf<T>() <= ReadBytesLeft);
+            if (Unsafe.SizeOf<T>() > ReadBytesLeft)
```

Comment:
This is opposite of how the I/O architecture currently works... If you look the read and write buffers, you'll see they (almost) never perform any bounds checking themselves. The assumption is that someone up the stack (the handler) has already performed the checks, and that if we're being asked to write X bytes, then there's at least that many bytes available in the buffer. This is why the check is currently in the form of a `Debug.Assert()` and not an if/throw.

The reason for this is performance. Doing a branching check on every little buffer I/O operation is pretty wasteful, when a single check could be done higher up the stack. For example, let's say some type handler needs to write 4 ints. It can do one check for 16 bytes (flushing if not available), and then write each of the 4 ints, rightfully assuming that there are 16 bytes of space. With this proposed chance, we'd be checking space on each one of the 4 ints.

Long story short, in my opinion bounds checking is the responsibility of the type handlers, not of the buffer classes.

Reply:
Strongly disagree. The results below shows that performance is the same. This is because you are not only writing ints, but also do heavy stuff like dispatching (choosing the right interface, calling the virtual method, etc.).

Reply:
@YohDeadfall it seems we're talking about several things in parallel, I'll try to separate them.

> Previously you have accessed the data using indexers and read it byte-by-byte. As you know, when you read or write to an array using the indexer, the JIT emits bounds check for each operation. Therefore, for int you had four checks. But nowdays there are no checks because of unsafe code, and it's very dangerous. Don't forget that it's a public API.

In my comment above I wasn't referring to JIT-generated array bounds checks, but rather about explicit code checks whether the buffer has enough capacity (`if (Unsafe.SizeOf<T>() > ReadBytesLeft)`). The way Npgsql is currently written, these happen in handlers rather than in the buffers.

If you're advocating for adding *additional* checks in buffer read/write methods (for added safety), maybe we can do that (see additional comments below). But we shouldn't *need* to: these would be *extra* safety checks. The specific bug you're trying to fix here is elsewhere.

> Strongly disagree. The results below shows that performance is the same. This is because you are not only writing ints, but also do heavy stuff like dispatching (choosing the right interface, calling the virtual method, etc.).

That's OK - the overhead of the extra branches *may* be dominated by heavier operations such as virtual method dispatching. But this could be the case in one scenario but not another (depending on the type being read/written?), or the weight of the virtualization may actually disappear once we (i.e. you :)) optimize things more, making the extra checks show up more in benchmarks.

I'll put it another way... Right now, Npgsql is architected in a certain way - buffer capacity checks are done up the stack in type handlers, and not down the stack in the buffers' primitive read/write methods. Whether this is important/good/significant isn't really relevant for fixing this bug - especially if we want to merge a fix to 4.0. I'm definitely not against reviewing everything for 4.1 (especially since we may have pipelines coming along...) but if there's a bug in 4.0 I'd like to fix that while keeping the current architecture.

Reply:
Yeah, I understand your position, but without that change there is no guarantee that everything works as expected. We don't have enough tests to cover potential errors like that. Moreover, it's very hard to determine the source problem in current situation (not for me, but for our users).

If you would like, we could backport only the `UnmappedTypeHandler` fix. It's a critical ug, so I would like to fix it as soon as possible.

Reply:
I get what you're saying, but I want to understand exactly where the current bug is coming from, and I still don't. I don't doubt that there's a bug somewhere :) Just trying to understand exactly what's going on. Here's how I understand things.

Aside from the buffer changes (which are extra protection), this PR adds a capacity check to `WriteWithLength<T>()` which it overrides from `NpgsqlTypeHandler<T>` - we check that there are at least 4 bytes left in the write buffer and flush if not. However, in `NpgsqlTypeHandler<T>`'s version of `WriteWithLength<T>()` you can see the exact same thing - the method *assumes* that the check was already done, and that there are 4 bytes available for the length. This assumption should be valid, because in the *current* code `WriteWithLength<T>()` is only meant to be called from `WriteWithLengthInternal<T>()`, which is where the check & flush happen.

Examining the code, the actual bug seems to be [here](https://github.com/npgsql/Npgsql/blob/e13f73f936440060a868ef4e73a17b7694f33856/src/Npgsql/TypeHandlers/UnmappedCompositeHandler.cs#L193) - when writing a composite as an object, `WriteWithLength()` is invoked directly without passing through `WriteWithLengthInternal()`. This would explain why the bug *only* occurs with composites, and not with all the other types (which inherit `WriteWithLength<T>()` from `NpgsqlTypeHandler<T>`, where no check & flush are present).

In other words, rather than making `WriteWithLength()` check & flush specifically for UnmappedCompositeHandler, we may need to simply correct its `WriteObjectWithLength()` it to invoke `WriteWithLengthInternal()` instead of `WriteWithLength()`... Does this make sense?

Then, if we want to add extra protection checks at the buffer level, let's do that in a separate PR etc., we can even decide to backport that to 4.0 as well (if there really isn't any perf impact). But if we do that, I think there may be other places that need this to be added.

Reply:
Yes, you are correct. This triggers the bug.

Okay, will remove checks here, but will introduce them as a part of handlers redesign. Is it fine?

Reply:
Thanks. I think it's OK to add the buffer-level checks, but how about doing it separately, before the handlers redesign? This would have two advantages:

* We can review and benchmark these checks separately from the more complex/architectural changes you are planning in the handlers redesign.
* We can decide to backport these extra checks to 4.0 (even the upcoming 4.0.4). If you do this as part of the handlers redesign that won't be possible.

What do you think?